src/client.rs

```diff
@@ -254,55 +310,212 @@ impl<T: Read + Write> Client<T> {
-            stream: BufStream::new(stream),
-            tag: INITIAL_TAG,
-            debug: false,
+            conn: Connection {
+                stream: BufStream::new(stream),
+                tag: INITIAL_TAG,
+                debug: false,
+            },
         }
     }
 
     /// Authenticate will authenticate with the server, using the authenticator given.
-    pub fn authenticate<A: Authenticator>(
-        &mut self,
+    pub fn authenticate<A: Authenticator> (
+        mut self,
         auth_type: &str,
         authenticator: A,
-    ) -> Result<()> {
-        self.run_command(&format!("AUTHENTICATE {}", auth_type))?;
+    ) -> ::std::result::Result<Session<T>, (Error, Client<T>)> {
+        // explicit match block neccessary to convert error to tuple and not bind self too early
+        // (see also comment on `login`)
+        ok_or_unauth_client_err!(self.run_command(&format!("AUTHENTICATE {}", auth_type)), self);
         self.do_auth_handshake(authenticator)
     }
 
     /// This func does the handshake process once the authenticate command is made.
-    fn do_auth_handshake<A: Authenticator>(&mut self, authenticator: A) -> Result<()> {
+    fn do_auth_handshake<A: Authenticator>(
+        mut self,
+        authenticator: A
+    ) -> ::std::result::Result<Session<T>, (Error, Client<T>)> {
         // TODO Clean up this code
         loop {
             let mut line = Vec::new();
-            self.readline(&mut line)?;
+            // explicit match blocks neccessary to convert error to tuple and not bind self too
+            // early (see also comment on `login`)
+            ok_or_unauth_client_err!(self.readline(&mut line), self);
 
             if line.starts_with(b"+") {
-                let data = parse_authenticate_response(String::from_utf8(line).unwrap())?;
+                let data = ok_or_unauth_client_err!(
+                    parse_authenticate_response(String::from_utf8(line).unwrap()), self);
                 let auth_response = authenticator.process(data);
 
-                self.write_line(auth_response.into_bytes().as_slice())?
+                ok_or_unauth_client_err!(self.write_line(auth_response.into_bytes().as_slice()), self);
             } else {
-                return self.read_response_onto(&mut line).map(|_| ());
+                ok_or_unauth_client_err!(self.read_response_onto(&mut line), self);
+                return Ok(Session { conn: self.conn });
             }
         }
     }
 
     /// Log in to the IMAP server.
-    pub fn login(&mut self, username: &str, password: &str) -> Result<()> {
-        self.run_command_and_check_ok(&format!(
-            "LOGIN {} {}",
-            validate_str(username)?,
-            validate_str(password)?
-        ))
+    pub fn login(
+        mut self,
+        username: &str,
+        password: &str
+    ) -> ::std::result::Result<Session<T>, (Error, Client<T>)> {
+        // note that we need the explicit match blocks here for two reasons:
```

Comment:
Should this comment now go away? Or perhaps be included in the macro?

Reply:
ah, yes.. next commit (also including the comments from above) soon-ish :-)